README.md

### DESCRIPTIONDESCRIPTION
This package is contain how to make idention.
identicon usually use as an avatar of your wallet / crypto avatar. it is unique because generated by your address. 

------------


#### METAMASKMETAMASK
![](https://i0.wp.com/www.alphr.com/wp-content/uploads/2022/04/Automatic-Token-Detection-on-a-PC-1.jpg?resize=670%2C675&ssl=1)

### Widget And JazziconData()
Add list of address wallet / coin. Make sure the address is valid. 

`Jazzicon.getIconWidget` need `JazziconData`. You can make JazziconData by calling 
`Jazzicon.getJazziconData()` 

Example : 
```dart
Jazzicon.getIconWidget(Jazzicon.getJazziconData(
160, 
address: "0xfe3526e15Bd6dAD0796B84bEb548b603B6B6E444"))
```

Result : 
![](https://github.com/kopianan/jazzicon/blob/main/assets/success.png?raw=true)


------------



If you make mistake and input the wrong code, example : 
**DO NOT DO THIS**
```dart
Jazzicon.getIconWidget(Jazzicon.getJazziconData(
160, 
address: "random string here"))
```
**ERROR RESULT**
![](https://github.com/kopianan/jazzicon/blob/main/assets/error.png?raw=true)

*note : if address is not valid the image will return grey image. *****
Example of valid address. You can open the [example](https://github.com/kopianan/jazzicon/blob/main/example/lib/main.dart "example") if you want to see more how to use this. 
```dart
  List<String> addresslist = [
    "0x123C3F273bE7d218A6Ab040D421b51237945ce11",
    "0x576C3F273bE7d218A6Eb040D421b51237945ce10",
    "0xEB8D9FDcf121bbae84c4E91aE47E9348c367c7F4",
    "0xb9bd886414b2d37b73c1266ebcb9d7f5fdb7383b",
    "0xb934F4294Df760290e969F355C107752400150bB",
    "0x318406407ea60F2E1C305bEd64Fb4A9db72AA6bD",
    "0x07Becc171101448476ABF27C1f73989F945A0742",
    "0xfe3526e15Bd6dAD0796B84bEb548b603B6B6E444",
  ];
```

------------


**Example** :
![](https://github.com/kopianan/jazzicon/blob/main/assets/full.png?raw=true)


------------

Open for contributor


CHANGELOG.md
## 1.1.0
* Add Custom Color
## 1.0.0
* Add documentation 
* Add error address, will return grey image if error. 

## 0.0.1

* TODO: Describe initial release.


lib/jazzicon.dart
library jazzicon;

import 'dart:math';

import 'package:flutter/material.dart';
import 'package:jazzicon/color_hsl.dart';
import 'package:jazzicon/jazziconshape.dart';
import 'package:jazzicon/mersennetwister19937.dart';

const List<Color> colors = [
  Color(0x0001888c), // teal             '#01888C', // teal
  Color(0x00fc7500), // bright orange    '#FC7500', // bright orange
  Color(0x00034f5d), // dark teal        '#034F5D', // dark teal
  Color(0x00f73f01), // orangered        '#F73F01', // orangered
  Color(0x00fc1960), // magenta          '#FC1960', // magenta
  Color(0x00c7144c), // raspberry        '#C7144C', // raspberry
  Color(0x00f3c100), // goldenrod        '#F3C100', // goldenrod
  Color(0x001598f2), // lightning blue   '#1598F2', // lightning blue
  Color(0x002465e1), // sail blue        '#2465E1', // sail blue
  Color(0x00f19e02), // gold             '#F19E02', // gold
];

const int wobble = 30;

const int shapeCount = 4;

class Jazzicon {
  ///Will return GREY only if the address is not valid.
  static JazziconData getJazziconData(double diameter,
      {int? seed, String? address, List<Color>? colorList}) {
    if (address != null && address.trim().isNotEmpty) {
      address = address.toLowerCase();

      if (address.startsWith("0x")) {
        try {
          String sub = address.substring(2, 10);
          seed = int.tryParse(sub, radix: 16);
          seed = MersenneTwister19937.unsigned32(seed! & 0xffffffff);
        } catch (e) {
          MersenneTwister19937 generator = MersenneTwister19937();

          List<JazziconShape> shapelist = [];
          for (var i = 0; i < 0; i++) {
            JazziconShape shape = _genShape(
                [Colors.black], diameter, i, shapeCount - 1, generator);

            shapelist.add(shape);
          }
          return JazziconData(
            size: diameter,
            background: Colors.grey,
            shapelist: shapelist,
          );
        }
      }
    }
    MersenneTwister19937 generator = MersenneTwister19937();
    generator.initGenrand(seed);

    List<Color> remainingColors = _hueShift(colorList ?? colors, generator);

    Color background = _genColor(remainingColors, generator);

    List<JazziconShape> shapelist = [];
    for (var i = 0; i < shapeCount - 1; i++) {
      JazziconShape shape =
          _genShape(remainingColors, diameter, i, shapeCount - 1, generator);

      shapelist.add(shape);
    }

    JazziconData jd = JazziconData(
        size: diameter, background: background, shapelist: shapelist);
    return jd;
  }

  static List<Color> _hueShift(
      List<Color> colors, MersenneTwister19937 generator) {
    double amount = (generator.random() * 30) - (wobble / 2);
    List<Color> _colors = colors.map((color) {
      return colorRotate(color, amount);
    }).toList();
    return _colors;
  }

  static Color _genColor(List<Color> colors, MersenneTwister19937 generator) {
    double r1 = generator.random();
    int idx = (colors.length * r1).floor();
    Color color = colors[idx];
    colors.removeAt(idx);
    return color;
  }

  static JazziconShape _genShape(List<Color> remainingColors, double diameter,
      int i, int total, MersenneTwister19937 generator) {
    double center = diameter / 2;

    double firstRot = generator.random();
    double angle = pi * 2 * firstRot;
    double velocity =
        diameter / total * generator.random() + (i * diameter / total);

    double tx = cos(angle) * velocity;
    double ty = sin(angle) * velocity;

    double secondRot = generator.random();
    double rot = (firstRot * 360) + secondRot * 180;
    Color fill = _genColor(remainingColors, generator);

    return JazziconShape(
        center: center,
        tx: tx,
        ty: ty,
        rotate: double.parse(rot.toStringAsFixed(1)),
        fill: fill);
  }

  static Widget getIconWidget(JazziconData jd, {double? size}) {
    double scale = 1;
    if (size != null) {
      scale = size / jd.size;
    }

    return StatefulBuilder(
      builder: (context, setState) {
        Widget view = ClipOval(
          child: Container(
            width: size ?? jd.size,
            height: size ?? jd.size,
            color: jd.background,
            child: Stack(
              children: jd.shapelist.map((shape) {
                Widget sp = Container(
                  width: size ?? jd.size,
                  height: size ?? jd.size,
                  color: shape.fill,
                );

                Widget spR = Transform.rotate(
                  angle: shape.rotate / 360 * 2 * pi,
                  // origin: Offset(shape.center, shape.center),
                  child: sp,
                );
                Widget spT = Transform.translate(
                  offset: Offset(shape.tx * scale, shape.ty * scale),
                  child: spR,
                );
                return Positioned(left: 0, right: 0, child: spT);
              }).toList(),
            ),
          ),
        );
        return view;
      },
    );
  }
}


lib/color_hsl.dart
import 'dart:math';
import 'dart:ui';

Map<String, double> colorToHSL(Color color) {
  // Convert hex to RGB first
  double r = color.red.toDouble();
  double g = color.green.toDouble();
  double b = color.blue.toDouble();
  // Then to HSL
  r /= 255;
  g /= 255;
  b /= 255;
  double cmin = min(min(r, g), b),
      cmax = max(max(r, g), b),
      delta = cmax - cmin,
      h = 0,
      s = 0,
      l = 0;

  if (delta == 0) {
    h = 0;
  } else if (cmax == r) {
    h = ((g - b) / delta) % 6;
  } else if (cmax == g) {
    h = (b - r) / delta + 2;
  } else {
    h = (r - g) / delta + 4;
  }

  // h = round(h * 60);
  h = (h * 60).roundToDouble();

  if (h < 0) h += 360;

  l = (cmax + cmin) / 2;
  s = delta == 0 ? 0 : delta / (1 - (2 * l - 1).abs());
  // s = +(s * 100).toFixed(1);
  s = double.parse((s * 100).toStringAsFixed(1));
  // l = +(l * 100).toFixed(1);
  l = double.parse((l * 100).toStringAsFixed(1));

  return {
    "h": h,
    "s": s,
    "l": l,
  };
}

Color colorRotate(Color color, double degrees) {
  Map<String, double> hsl = colorToHSL(color);
  double hue = hsl["h"] ?? 0;
  hue = (hue + degrees) % 360;
  hue = hue < 0 ? 360 + hue : hue;
  hsl["h"] = hue;
  double h = hsl["h"] ?? 0;
  double s = hsl["s"] ?? 0;
  double l = hsl["l"] ?? 0;

  s /= 100;
  l /= 100;

  double c = (1 - (2 * l - 1).abs()) * s,
      x = c * (1 - ((h / 60) % 2 - 1).abs()),
      m = l - c / 2,
      r = 0,
      g = 0,
      b = 0;

  if (0 <= h && h < 60) {
    r = c;
    g = x;
    b = 0;
  } else if (60 <= h && h < 120) {
    r = x;
    g = c;
    b = 0;
  } else if (120 <= h && h < 180) {
    r = 0;
    g = c;
    b = x;
  } else if (180 <= h && h < 240) {
    r = 0;
    g = x;
    b = c;
  } else if (240 <= h && h < 300) {
    r = x;
    g = 0;
    b = c;
  } else if (300 <= h && h < 360) {
    r = c;
    g = 0;
    b = x;
  }
  // Having obtained RGB, convert channels to hex
  int _r = ((r + m) * 255).round();
  int _g = ((g + m) * 255).round();
  int _b = ((b + m) * 255).round();
  return Color.fromARGB(0xff, _r, _g, _b);
}


lib/mersennetwister19937.dart
class MersenneTwister19937 {
  /* Period parameters */
  //c//#define N 624
  //c//#define M 397
  //c//#define MATRIX_A 0x9908b0dfUL   /* constant vector a */
  //c//#define UPPER_MASK 0x80000000UL /* most significant w-r bits */
  //c//#define LOWER_MASK 0x7fffffffUL /* least significant r bits */
  static const int N = 624;
  static const int M = 397;
  static const int matrixA = 0x9908b0df; /* constant vector a */
  static const int upperMask = 0x80000000; /* most significant w-r bits */
  static const int lowerMask = 0x7fffffff; /* least significant r bits */
  //c//static unsigned long mt[N]; /* the array for the state vector  */
  //c//static int mti=N+1; /* mti==N+1 means mt[N] is not initialized */
  late List<int>
      mt; // = new List<num>(N);   /* the array for the state vector  */
  late int
      mti; // = N+1;           /* mti==N+1 means mt[N] is not initialized */

  MersenneTwister19937() {
    mt = List.filled(N, 0);
    mti = N + 1;
  }

  static unsigned32(
      int n1) // returns a 32-bits unsiged integer from an operand to which applied a bit operator.
  {
    return n1 < 0 ? (n1 ^ upperMask) + upperMask : n1;
  }

  subtraction32(int n1,
      int n2) // emulates lowerflow of a c 32-bits unsiged integer variable, instead of the operator -. these both arguments must be non-negative integers expressible using unsigned 32 bits.
  {
    return n1 < n2
        ? unsigned32((0x100000000 - (n2 - n1)) & 0xffffffff)
        : n1 - n2;
  }

  addition32(int n1,
      int n2) // emulates overflow of a c 32-bits unsiged integer variable, instead of the operator +. these both arguments must be non-negative integers expressible using unsigned 32 bits.
  {
    return unsigned32((n1 + n2) & 0xffffffff);
  }

  multiplication32(int n1,
      int n2) // emulates overflow of a c 32-bits unsiged integer variable, instead of the operator *. these both arguments must be non-negative integers expressible using unsigned 32 bits.
  {
    var sum = 0;
    for (var i = 0; i < 32; ++i) {
      if (((n1 >> i) & 0x1) > 0) {
        sum = addition32(sum, unsigned32(n2 << i));
      }
    }
    return sum;
  }

  /* initializes mt[N] with a seed */
  //c//void init_genrand(unsigned long s)
  void initGenrand(int? s) {
    s = s ?? DateTime.now().millisecond;

    //c//mt[0]= s & 0xffffffff;
    mt[0] = unsigned32(s & 0xffffffff);
    for (mti = 1; mti < N; mti++) {
      mt[mti] =
          //c//(1812433253 * (mt[mti-1] ^ (mt[mti-1] >> 30)) + mti);
          addition32(
              multiplication32(
                  1812433253, unsigned32(mt[mti - 1] ^ (mt[mti - 1] >> 30))),
              mti);
      /* See Knuth TAOCP Vol2. 3rd Ed. P.106 for multiplier. */
      /* In the previous versions, MSBs of the seed affect   */
      /* only MSBs of the array mt[].                        */
      /* 2002/01/09 modified by Makoto Matsumoto             */
      //c//mt[mti] &= 0xffffffff;
      mt[mti] = unsigned32(mt[mti] & 0xffffffff);
      // print("init_seed  index=$mti  ${mt[mti]}");
      /* for >32 bit machines */
    }
  }

  /* initialize by an array with array-length */
  /* init_key is the array for initializing keys */
  /* key_length is its length */
  /* slight change for C++, 2004/2/26 */
  //c//void init_by_array(unsigned long init_key[], int key_length)
  void initByArray(var initKey, var keyLength) {
    //c//int i, j, k;
    late int i;
    late int j;
    late dynamic k;
    //c//init_genrand(19650218);
    initGenrand(19650218);
    i = 1;
    j = 0;
    k = (N > keyLength ? N : keyLength);
    for (; k; k--) {
      //c//mt[i] = (mt[i] ^ ((mt[i-1] ^ (mt[i-1] >> 30)) * 1664525))
      //c// + init_key[j] + j; /* non linear */
      mt[i] = addition32(
          addition32(
              unsigned32(mt[i] ^
                  multiplication32(
                      unsigned32(mt[i - 1] ^ (mt[i - 1] >> 30)), 1664525)),
              initKey[j]),
          j);
      mt[i] =
          //c//mt[i] &= 0xffffffff; /* for WORDSIZE > 32 machines */
          unsigned32(mt[i] & 0xffffffff);
      i++;
      j++;
      if (i >= N) {
        mt[0] = mt[N - 1];
        i = 1;
      }
      if (j >= keyLength) j = 0;
    }
    for (k = N - 1; k; k--) {
      //c//mt[i] = (mt[i] ^ ((mt[i-1] ^ (mt[i-1] >> 30)) * 1566083941))
      //c//- i; /* non linear */
      mt[i] = subtraction32(
          unsigned32(mt[i] ^
              multiplication32(
                  unsigned32(mt[i - 1] ^ (mt[i - 1] >> 30)), 1566083941)),
          i);
      //c//mt[i] &= 0xffffffff; /* for WORDSIZE > 32 machines */
      mt[i] = unsigned32(mt[i] & 0xffffffff);
      i++;
      if (i >= N) {
        mt[0] = mt[N - 1];
        i = 1;
      }
    }
    mt[0] = 0x80000000; /* MSB is 1; assuring non-zero initial array */
  }

  /* generates a random number on [0,0xffffffff]-interval */
  //c//unsigned long genrand_int32(void)
  int genrandInt32() {
    //c//unsigned long y;
    //c//static unsigned long mag01[2]={0x0UL, MATRIX_A};
    int y = 0;
    List<int> mag01 = []; //= new Array(0x0, MATRIX_A);
    mag01.add(0x0);
    mag01.add(matrixA);
    /* mag01[x] = x * MATRIX_A  for x=0,1 */

    if (mti >= N) {
      /* generate N words at one time */
      //c//int kk;
      int kk;

      if (mti == N + 1) {
        initGenrand(5489); /* a default initial seed is used */
      }

      for (kk = 0; kk < N - M; kk++) {
        //c//y = (mt[kk]&UPPER_MASK)|(mt[kk+1]&LOWER_MASK);
        //c//mt[kk] = mt[kk+M] ^ (y >> 1) ^ mag01[y & 0x1];
        y = unsigned32((mt[kk] & upperMask) | (mt[kk + 1] & lowerMask));
        mt[kk] = unsigned32(mt[kk + M] ^ (y >> 1) ^ mag01[y & 0x1]);
      }
      for (; kk < N - 1; kk++) {
        //c//y = (mt[kk]&UPPER_MASK)|(mt[kk+1]&LOWER_MASK);
        //c//mt[kk] = mt[kk+(M-N)] ^ (y >> 1) ^ mag01[y & 0x1];
        y = unsigned32((mt[kk] & upperMask) | (mt[kk + 1] & lowerMask));
        mt[kk] = unsigned32(mt[kk + (M - N)] ^ (y >> 1) ^ mag01[y & 0x1]);
      }
      //c//y = (mt[N-1]&UPPER_MASK)|(mt[0]&LOWER_MASK);
      //c//mt[N-1] = mt[M-1] ^ (y >> 1) ^ mag01[y & 0x1];
      y = unsigned32((mt[N - 1] & upperMask) | (mt[0] & lowerMask));
      mt[N - 1] = unsigned32(mt[M - 1] ^ (y >> 1) ^ mag01[y & 0x1]);
      mti = 0;
    }

    y = mt[mti++];

    /* Tempering */
    //c//y ^= (y >> 11);
    //c//y ^= (y << 7) & 0x9d2c5680;
    //c//y ^= (y << 15) & 0xefc60000;
    //c//y ^= (y >> 18);
    y = unsigned32(y ^ (y >> 11));
    y = unsigned32(y ^ ((y << 7) & 0x9d2c5680));
    y = unsigned32(y ^ ((y << 15) & 0xefc60000));
    y = unsigned32(y ^ (y >> 18));

    return y;
  }

  /* generates a random number on [0,0x7fffffff]-interval */
  //c//long genrand_int31(void)
  int genrandInt31() {
    //c//return (genrand_int32()>>1);
    return (genrandInt32() >> 1);
  }

  /* generates a random number on [0,1]-real-interval */
  //c//double genrand_real1(void)
  double genrandReal1() {
    //c//return genrand_int32()*(1.0/4294967295.0);
    return genrandInt32() * (1.0 / 4294967295.0);
    /* divided by 2^32-1 */
  }

  /* generates a random number on [0,1)-real-interval */
  //c//double genrand_real2(void)
  double genrandReal2() {
    //c//return genrand_int32()*(1.0/4294967296.0);
    return genrandInt32() * (1.0 / 4294967296.0);
    /* divided by 2^32 */
  }

  /* generates a random number on (0,1)-real-interval */
  //c//double genrand_real3(void)
  double genrandReal3() {
    //c//return ((genrand_int32()) + 0.5)*(1.0/4294967296.0);
    return ((genrandInt32()) + 0.5) * (1.0 / 4294967296.0);
    /* divided by 2^32 */
  }

  /* generates a random number on [0,1) with 53-bit resolution*/
  //c//double genrand_res53(void)
  double genrandRes53() {
    //c//unsigned long a=genrand_int32()>>5, b=genrand_int32()>>6;
    var a = genrandInt32() >> 5, b = genrandInt32() >> 6;
    return (a * 67108864.0 + b) * (1.0 / 9007199254740992.0);
  }
/* These real versions are due to Isaku Wada, 2002/01/09 added */

  double random() {
    return genrandInt32() * (1.0 / 4294967296.0);
    /* divided by 2^32 */
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:jazzicon/jazziconshape.dart';
import 'package:jazzicon/jazzicon.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  JazziconData? jd;

  List<String> addresslist = [
    "0x123C3F273bE7d218A6Ab040D421b51237945ce11",
    "0x576C3F273bE7d218A6Eb040D421b51237945ce10",
    "0xEB8D9FDcf121bbae84c4E91aE47E9348c367c7F4",
    "0xb9bd886414b2d37b73c1266ebcb9d7f5fdb7383b",
    "0xb934F4294Df760290e969F355C107752400150bB",
    "0x318406407ea60F2E1C305bEd64Fb4A9db72AA6bD",
    "0x07Becc171101448476ABF27C1f73989F945A0742",
    "0xfe3526e15Bd6dAD0796B84bEb548b603B6B6E444",
  ];

  List<JazziconData> jdlist = [];

  @override
  void initState() {
    super.initState();
    // jd = Jazzicon.getJazziconData(200, address: "0x85CddC7f65410E9Cb94D959Ac57405ee0EcBFBE3");
    for (var address in addresslist) {
      jdlist.add(Jazzicon.getJazziconData(160, address: address));
    }
  }

  void _incrementCounter() {
    setState(() {});
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        child: Wrap(
          spacing: 20,
          runSpacing: 20,
          children: jdlist.map((e) {
            return Jazzicon.getIconWidget(e);
          }).toList(),
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
